File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var diameterOfBinaryTree = function(root) {
    let maxD = 0
    function dfs(node) {
        if(node === null) return -1

        const lh = 1 + dfs(node.left)
        const rh = 1 + dfs(node.right)

        maxD = Math.max(maxD, lh+rh)

        return Math.max(lh, rh)
    }
    dfs(root)
    return maxD
};

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    diameter = 0

    def dfs(self, root: Optional[TreeNode]) -> int:
        if root == None:
            return 0
        
        lHeight = self.dfs(root.left)
        rHeight = self.dfs(root.right)
        self.diameter = max(self.diameter, lHeight+rHeight)
        return 1 + max(lHeight, rHeight)

    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.dfs(root)
        return self.diameter
        